### ISSUE TITLE
[3.0] [Pages] [Crash] NullPointerException while serializing not yet created pages

### ISSUE THREAD

**Issue by Lastaapps** [arkivanov/Decompose#612]
Hi, thank you for a great project. I recently migrated my project to Decompose and noticed crash on configuration change while using the pages navigation.

- version: 3.0.0-alpha04

The problem is, that default state of `ChildItem.Destroyed` is `null`. If a child in pages navigation was not created yet the `null` value remains in the `savedState` field. On a configuration change, the engine tries to serialize it, but `null` cannot be serialized by the current implementation of `SerializableContainer.Serializable` (and it should stay that way, I think).

I think this issue should also affect the stack navigation, but over there every component is naturally created at least once (maybe stack created from a deep link could be an issue).

Once I go through all the pages and let then create at least once, `null` is replaced by a `SerializableContainer` with an empty map inside and everything works as expected.

[My project](https://github.com/Lastaapps/menza) (not a minimal example, but not complex either).

Proposed solutions:
- use some kind of serializable default value instead of `null`
- make `SerializableContainer.Serializer` support `null` values

Thanks

```
java.lang.NullPointerException: Parameter specified as non-null is null: method com.arkivanov.essenty.statekeeper.SerializableContainer$Serializer.serialize, parameter value
at com.arkivanov.essenty.statekeeper.SerializableContainer$Serializer.serialize(Unknown Source:8)
at com.arkivanov.essenty.statekeeper.SerializableContainer$Serializer.serialize(SerializableContainer.kt:62)
at kotlinx.serialization.json.internal.StreamingJsonEncoder.encodeSerializableValue(StreamingJsonEncoder.kt:226)
at kotlinx.serialization.encoding.AbstractEncoder.encodeSerializableElement(AbstractEncoder.kt:80)
at kotlinx.serialization.internal.CollectionLikeSerializer.serialize(CollectionSerializers.kt:69)
at kotlinx.serialization.json.internal.StreamingJsonEncoder.encodeSerializableValue(StreamingJsonEncoder.kt:226)
at kotlinx.serialization.encoding.AbstractEncoder.encodeSerializableElement(AbstractEncoder.kt:80)
at com.arkivanov.decompose.router.children.SavedState.write$Self$decompose_release(ChildrenFactory.kt:191)
at com.arkivanov.decompose.router.children.SavedState$$serializer.serialize(ChildrenFactory.kt:191)
at com.arkivanov.decompose.router.children.SavedState$$serializer.serialize(ChildrenFactory.kt:191)
...
```

EDIT: added stack trace

**Comment by arkivanov** [comment]
`SerializableContainer` is able to serialize `null` values and it's covered with [tests](https://github.com/arkivanov/Essenty/blob/master/state-keeper/src/commonTest/kotlin/com/arkivanov/essenty/statekeeper/SerializableContainerTest.kt). There are also tests for Child Pages state restoration and it [covers](https://github.com/arkivanov/Decompose/blob/f52ed09e5f82ce229e4e0104d75ff167207de893/decompose/src/commonTest/kotlin/com/arkivanov/decompose/router/pages/ChildPagesSavedStateTest.kt#L22) not-yet-created children as well. Would you be able to provide a minimal reproducer or a runnable test?

**Comment by Lastaapps** [comment]
Sure, here you go. I don't think I can go smaller then this. Basic template in AS + decomponse 3.0.0-alpha04

- project: https://github.com/Lastaapps/decomposepageserialization
- code: https://github.com/Lastaapps/decomposepageserialization/blob/main/app/src/main/java/cz/lastaapps/decompose_page_serialization/MainActivity.kt

```
Process: cz.lastaapps.decompose_page_serialization, PID: 18780
java.lang.NullPointerException: Parameter specified as non-null is null: method com.arkivanov.essenty.statekeeper.SerializableContainer$Serializer.serialize, parameter value
at com.arkivanov.essenty.statekeeper.SerializableContainer$Serializer.serialize(Unknown Source:8)
at com.arkivanov.essenty.statekeeper.SerializableContainer$Serializer.serialize(SerializableContainer.kt:62)
at kotlinx.serialization.json.internal.StreamingJsonEncoder.encodeSerializableValue(StreamingJsonEncoder.kt:226)
at kotlinx.serialization.encoding.AbstractEncoder.encodeSerializableElement(AbstractEncoder.kt:80)
at kotlinx.serialization.internal.CollectionLikeSerializer.serialize(CollectionSerializers.kt:69)
at kotlinx.serialization.json.internal.StreamingJsonEncoder.encodeSerializableValue(StreamingJsonEncoder.kt:226)
at kotlinx.serialization.encoding.AbstractEncoder.encodeSerializableElement(AbstractEncoder.kt:80)
at com.arkivanov.decompose.router.children.SavedState.write$Self$decompose_release(ChildrenFactory.kt:191)
at com.arkivanov.decompose.router.children.SavedState$$serializer.serialize(ChildrenFactory.kt:191)
at com.arkivanov.decompose.router.children.SavedState$$serializer.serialize(ChildrenFactory.kt:191)
at kotlinx.serialization.json.internal.StreamingJsonEncoder.encodeSerializableValue(StreamingJsonEncoder.kt:226)
at kotlinx.serialization.json.internal.JsonStreamsKt.encodeByWriter(JsonStreams.kt:31)
at kotlinx.serialization.json.JvmStreamsKt.encodeToStream(JvmStreams.kt:25)
at com.arkivanov.essenty.statekeeper.Utils_javaKt.serialize(Utils.java.kt:22)
at com.arkivanov.essenty.statekeeper.SerializableContainer$Serializer.serialize(SerializableContainer.kt:71)
at com.arkivanov.essenty.statekeeper.SerializableContainer$Serializer.serialize(SerializableContainer.kt:67)
at com.arkivanov.essenty.statekeeper.SerializableContainer$Serializer.serialize(SerializableContainer.kt:62)
at kotlinx.serialization.json.internal.StreamingJsonEncoder.encodeSerializableValue(StreamingJsonEncoder.kt:226)
at kotlinx.serialization.encoding.AbstractEncoder.encodeSerializableElement(AbstractEncoder.kt:80)
at kotlinx.serialization.internal.MapLikeSerializer.serialize(CollectionSerializers.kt:123)
at kotlinx.serialization.json.internal.StreamingJsonEncoder.encodeSerializableValue(StreamingJsonEncoder.kt:226)
at kotlinx.serialization.encoding.AbstractEncoder.encodeSerializableElement(AbstractEncoder.kt:80)
at com.arkivanov.essenty.statekeeper.DefaultStateKeeperDispatcher$SavedState.write$Self$state_keeper_release(DefaultStateKeeperDispatcher.kt:52)
at com.arkivanov.essenty.statekeeper.DefaultStateKeeperDispatcher$SavedState$$serializer.serialize(DefaultStateKeeperDispatcher.kt:52)
at com.arkivanov.essenty.statekeeper.DefaultStateKeeperDispatcher$SavedState$$serializer.serialize(DefaultStateKeeperDispatcher.kt:52)
at kotlinx.serialization.json.internal.StreamingJsonEncoder.encodeSerializableValue(StreamingJsonEncoder.kt:226)
at kotlinx.serialization.json.internal.JsonStreamsKt.encodeByWriter(JsonStreams.kt:31)
at kotlinx.serialization.json.JvmStreamsKt.encodeToStream(JvmStreams.kt:25)
at com.arkivanov.essenty.statekeeper.Utils_javaKt.serialize(Utils.java.kt:22)
at com.arkivanov.essenty.statekeeper.SerializableContainer$Serializer.serialize(SerializableContainer.kt:71)
at com.arkivanov.essenty.statekeeper.SerializableContainer$Serializer.serialize(SerializableContainer.kt:67)
at com.arkivanov.essenty.statekeeper.SerializableContainer$Serializer.serialize(SerializableContainer.kt:62)
at kotlinx.serialization.json.internal.StreamingJsonEncoder.encodeSerializableValue(StreamingJsonEncoder.kt:226)
at kotlinx.serialization.json.internal.JsonStreamsKt.encodeByWriter(JsonStreams.kt:31)
at kotlinx.serialization.json.JvmStreamsKt.encodeToStream(JvmStreams.kt:25)
at com.arkivanov.essenty.statekeeper.Utils_javaKt.serialize(Utils.java.kt:22)
at com.arkivanov.essenty.statekeeper.AndroidExtKt.StateKeeper$lambda$2(AndroidExt.kt:38)
at com.arkivanov.essenty.statekeeper.AndroidExtKt.$r8$lambda$rH3LitWBDg4vgeZlo5vRdXkobCo(Unknown Source:0)
at com.arkivanov.essenty.statekeeper.AndroidExtKt$$ExternalSyntheticLambda0.saveState(D8$$SyntheticClass:0)
at androidx.savedstate.SavedStateRegistry.performSave(SavedStateRegistry.kt:247)
at androidx.savedstate.SavedStateRegistryController.performSave(SavedStateRegistryController.kt:81)
at androidx.activity.ComponentActivity.onSaveInstanceState(ComponentActivity.java:374)
at android.app.Activity.performSaveInstanceState(Activity.java:2349)
at android.app.Instrumentation.callActivityOnSaveInstanceState(Instrumentation.java:1635)
at android.app.ActivityThread.callActivityOnSaveInstanceState(ActivityThread.java:6509)
at android.app.ActivityThread.callActivityOnStop(ActivityThread.java:5911)
at android.app.ActivityThread.handleRelaunchActivityInner(ActivityThread.java:6467)
at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:6377)
at android.app.servertransaction.ActivityRelaunchItem.execute(ActivityRelaunchItem.java:76)
at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:139)
at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:96)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2685)
at android.os.Handler.dispatchMessage(Handler.java:106)
at android.os.Looper.loopOnce(Looper.java:230)
at android.os.Looper.loop(Looper.java:319)
at android.app.ActivityThread.main(ActivityThread.java:8913)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:608)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1103)
```

**Comment by Lastaapps** [comment]
> `SerializableContainer` is able to serialize null values

Also, this statement is true, it can serialize null stored in `data`/`holder`, the problem occurs when the `SerializableContainer` itself is `null`. The issue comes from `fun saveChildState(): List<SerializableContainer?>` getting serialized later.

**Comment by arkivanov** [comment]
Thanks, reproduced! I will fix this in the next release.

**Comment by arkivanov** [comment]
> The issue comes from fun saveChildState(): List<SerializableContainer?> getting serialized later.

Correct. There is a bug in `SerializableContainer.Serializable`. The reason why this was never cached by tests is because the actual serialization is not performed, the same instance of `SerializableContainer` is passed back as saved state. I will need to fix tests as well.

**Comment by Lastaapps** [comment]
Nice, thank you for your great work again.

**Comment by arkivanov** [comment]
Fixed in Decompose [3.0.0-alpha05](https://github.com/arkivanov/Decompose/releases/tag/3.0.0-alpha05).
